pkg/api/meta/scheme_test.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package meta_test
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// These types do not follow the list convention as documented in
+// docs/devel/api-convention.md
+var listTypeExceptions = sets.NewString("APIGroupList", "APIResourceList")
+
+func validateListType(target reflect.Type) error {
+	// exceptions
+	if listTypeExceptions.Has(target.Name()) {
+		return nil
+	}
+	if strings.HasSuffix(target.Name(), "List") {
+		var foundMetadata, foundItems bool
+		for i := 0; i < target.NumField(); i++ {
+			field := target.Field(i)
+			tag := field.Tag.Get("json")
+			switch {
+			case strings.HasPrefix(tag, "metadata"):
+				foundMetadata = true
+			case tag == "items":
+				foundItems = true
+				if field.Type.Kind() != reflect.Slice {
+					return fmt.Errorf("Expected items to be slice, got %s", field.Type.Kind())
+				}
+			}
+		}
+		if !foundMetadata {
+			return fmt.Errorf("Expected type %s to contain \"metadata\"", target.Name())
+		}
+		if !foundItems {
+			return fmt.Errorf("Expected type %s to contain \"items\"", target.Name())
+		}
+	} else {
+		// if a type contains field Items with JSON tag "items", its name should end with List.
+		for i := 0; i < target.NumField(); i++ {
+			if target.Field(i).Name == "Items" && target.Field(i).Tag.Get("json") == "items" {
+				return fmt.Errorf("Type %s has Items, its name is expected to end with \"List\"", target.Name())
+			}
+		}
+	}
+	return nil
+}
+
+// TestListTypes verifies that no external type violates the api convention of
+// list types.
+func TestListTypes(t *testing.T) {
+	for groupKey, group := range testapi.Groups {
+		for kind, target := range group.ExternalTypes() {
```

Comment:
That's fine - eventually serialization test needs to get hoisted out of pkg/api anyway.